### ISSUE TITLE
Dynamic registry

### ISSUE THREAD

**Issue by nowox** [hgrecco/pint#528]
Pint is an awesome package, perhaps not popular enough to be on [Awesome Python](https://github.com/vinta/awesome-python) or [Anaconda](https://docs.continuum.io/anaconda/pkg-docs). 

I have a special request that would make Pint very useful for my needs. I work with motion control applications and our electronics carry units like position, current, voltage, force and velocity. 

Communication with our electronics is made through low-level registers that use internal units that can be later converted to SI units using a conversion factor.

This conversion factor depends on some settings such as the resolution of the position sensor or the type of actuator used. It can be dynamically modified if the setting is changed by the user. 

I would like to use Pint to take care of all these units unfortunately there is no way I can work with different registry together neither update the units definition at runtime. 

To summary, I would to be able to do this:

```
def RegistryBuilder(units):
    ureg = UnitRegistry()
    for k, v in units.items():
        ureg.define('%s = %s' % (k,v))
    return ureg
 
class Device(object):
    _conv_factor = {}

    def __init__(self): 
        self.ureg = RegistryBuilder(self._conv_factors)

    @property
    def position(self):
        return self._internal_position * self.ureg.u

    @property
    def current(self):
        return self._internal_current * self.ureg.v

class DeviceA(Device):
    _conv_factors = {'u': '0.00123 meters', 'v': '12.0 amperes'}

class DeviceB(Device):
    _conv_factors = {'u': '0.00123 meters', 'v': '6.0 amperes'}

a = DeviceA()
b = DeviceB()

a.position + b.position # Valid because they both carry the same conversion unit
a.current + b.current # Invalid because the internal conversion factor differ

v = a.ureg.Quantity('1 ampere') + b.ureg.Quantity('1 ampere') # Valid
v.to('v') # Invalid

```

How much work would be required to support this kind of features which can make all the difference in my application?

In other words I would like to be able to work with different registry only if: 

- The base unit definition is the same for both registries

**Comment by hgrecco** [comment]
> Pint is an awesome package,

Thanks for the support!

I think that the right way is not to expose the calibration outside the device. Something like:
```python
import pint

ureg = pint.UnitRegistry()

class Device:

    def __init__(position_cal, current_cal):
        self.position_cal = position_cal
        self.current_cal = current_cal

    @property
    def position(self):
        return self._internal_position * self.position_cal

    @property
    def current(self):
        return self._internal_current * self.current_cal

a = Device(0.00123 * ureg.meter, 12.0 * ureg.amperes)
b = Device(0.00123 * ureg.meter, 6.0 * ureg.amperes)
```
or defining a class as you suggested:
```python
import pint

ureg = pint.UnitRegistry()

class Device:

    @property
    def position(self):
        return self._internal_position * self._conv_factors['position_cal']

    @property
    def current(self):
        return self._internal_current * self._conv_factors['current_cal']

class DeviceA(Device):
    _conv_factors = {'position_cal': 0.00123 * ureg.meter, 'current_cal': 12.0 * ureg.amperes}

class DeviceB(Device):
    _conv_factors = {'position_cal': 0.00123 * ureg.meter, 'current_cal':  6.0 * ureg.amperes}

a = DeviceA()
b = DeviceB()
```

I am not sure if this fullfills all your needs but is really clean.

**Comment by nowox** [comment]
Nice suggestion, this is exactly what I have implementer so far. Unfortunately it does not fulfills all requirements. 

A device has more than 50 different *internal units* definitions. I would like to access these internal units also from the outside: 

```
a = DeviceA()
a.position.to('meters')
a.position.to('internal_meters')
```

So the initial question is still what I am looking for.

**Comment by hgrecco** [comment]
Closing. Feel free to reopen. And look at the new contexts with the ability to redefine units.
